pycqed/instrument_drivers/physical_instruments/ZurichInstruments/ZI_HDAWG8.py

```diff
@@ -1,417 +1,102 @@
+"""
+Notes:
+- this is the application dependent part of the ZI HDAWG driver: it handles all codeword related
+  functionality including DIO configuration. Generic parts reside in ZI_HDAWG_core.py
+
+To do:
+- replace print() by logging
+-
+
+Changelog:
+
+20190206 WJV
+- started this Changelog
+- manually checked against diverted branch HDAWG_V2_Verification:
+    - the following functions match:
+        - _find_valid_delays
+        - _set_dio_delay
+        - ensure_symmetric_strobe
+        - calibrate_dio_protocol(self, awgs_and_sequences, verbose=False)
+        - _get_edges
+        - _is_dio_strb_symmetric
+        - _analyze_dio_data
+    - the following were already commented out here:
+        - _check_protocol
+        - _print_check_protocol_error_message
+        - calibrate_dio
+        - calibrate_dio_protocol(self)
+    So we conclude all relevant changes of HDAWG_V2_Verification made it here,
+    albeit in a different order that clutters the diff.
+- removed the above mentioned 4 functions that were commented out
+- added comments, organized code into sections
+- made some functions 'private'
+- NB: none of the above should change anything for real
+- moved enabling of outputs to end in configure_codeword_protocol
+
+20190207 WJV
+- added assure_ext_clock()
+
+20190212 WJV
+- separated off application independent stuff into ZI_HDAWG_core class, this
+  file will keep application dependent stuff
+- addressed many warnings identified by PyCharm
+
+20190214 WJV
+- added activate_new_dio_triggering()
+- moved in _add_extra_parameters() and _add_codeword_parameters()
+- moved out _set_dio_delay()
+
+20190417 WJV
+- merged branch 'develop' into 'feature/cc'
+
+20190429 WJV
+- merged branch 'QCC_testing' into 'feature/cc', changes:
+    upload_waveform_realtime was updated, moved it to ZI_HDAWG_core.py again
+
+20190618 WJV
+- merged branch 'develop' into 'feature/cc', changes:
+    upload_waveform_realtime was updated, moved it to ZI_HDAWG_core.py again
+
+
+"""
+
 import time
 import logging
-import os
 import numpy as np
-from . import zishell_NH as zs
+
+from .ZI_HDAWG_core import ZI_HDAWG_core
 from qcodes.utils import validators as vals
-from .ZI_base_instrument import ZI_base_instrument
 from qcodes.instrument.parameter import ManualParameter
-from zlib import crc32
-
-import ctypes
-from ctypes.wintypes import MAX_PATH
 
+log = logging.getLogger(__name__)
 
-class ZI_HDAWG8(ZI_base_instrument):
-    """
-    This is PycQED/QCoDeS driver driver for the Zurich Instruments HD AWG-8.
-
-    Parameter files are generated from the python API of the instrument
-    using the "create_parameter_file" method in the ZI_base_instrument class.
-    These are used to add parameters to the instrument.
-    """
+class ZI_HDAWG8(ZI_HDAWG_core):
```

Comment:
Why different inheritance?

Reply:
I moved generic functionality (i.e. stuff not related to how we use codewords) of the AWG8 to a new class ZI_HDAWG_core. This provides a - IMHO much wanted - separation of functionality: basically DCL should never have to update that class, and ZI should never have to update ZI_HDAWG8